### ISSUE TITLE
Mechanism to checksum assembly-compilation to determine if it has changed or not

### ISSUE THREAD

**Issue by tapika** [icsharpcode/ILSpy#2220]
I have raised same issue on dotnet / core, ticket 5577.

But was wondering if it's possible to support the same functionality using ilspy , reusable functionality e.g. using ICSharpCode.Decompiler. Please note that I don't need full decompilation support, but as quick as possible mechanism to checksum assembly. See full description.

It should be possible to calculate hash or checksum over compiled assembly, so same input files would produce the same checksum.

This is useful to identify whether assembly has changed or not - and act upon that - for example execute correspondent tests or release nuget package.

Doing something like this was attempted in several ways - for example using ildasm:

1. https://www.codeproject.com/Articles/501631/Compare-Two-DLL-Files-Programmatically-Using-Hash
dumpbin:

2. https://stackoverflow.com/questions/3316505/how-can-i-know-if-an-assembly-really-did-change
mono.Cecil - programmatically:

3. https://github.com/jbevain/cecil/pull/618/files#diff-0e0bdb24805aa1a95d31f9d8561d21b80a400756610cf1d20e4392359589e2ef
GetBuildIndependentHash.

There also exists an option for Roslyn to use deterministric builds:

4. https://blog.paranoidcoding.com/2016/04/05/deterministic-builds-in-roslyn.html
To my best understanding using approach 1 or 2 is slow, because it's call to external tool. Which also might or might not work depending on version of compiler and / or tool.

Brief analysis of 4 - showed that timestamp still changes within a compilation. Suspect if I'll sign assembly - it will change even more. 3 could be working solution, but my brief trials showed the differences to appear in a middle of byte code (.text section, as a platform I have used .net framework).

.net should provide mechanism / api to be able to reliably checksum assembly to identify whether it changed or not.

Checksuming should be able to exclude:

- Assembly signature
- Build timestamp ('Time-date stamp:' in ildasm)
- GUID of debug data ('MVID' in ildasm ?)
- 'Image base' is ildasm
- Anything else that changes from build to build.

It should be possible to compile same input sources on two different machines, and then run through assembly checksum and get identical checksum after that.

**Comment by dgrunwald** [comment]
The `/deterministic` switch of the C# compiler should do what you want. If it doesn't, that's a bug in the C# compiler.

Sure, you can alternatively use ILSpy to decompile the assembly and compare the decompiled code.
But I don't see what kind of change to ILSpy you are requesting here?

**Comment by tapika** [comment]
What i have tested - /deterministic does not work correctly. Also Im not sure how it will work with signing - timestamp should change anyway ?! 

I would like to get some sort of API from ilspy base assemblies. Full reverse engineering is not needed in this case, only load parts of assembly to be able determine checksum.

I guess ilspy is closest utility towards what I want to accomplish.

**Comment by dgrunwald** [comment]
The timestamp shouldn't change with `/deterministic`. You'll first need to figure out which portion of the assembly is changing unexpectedly. Otherwise it's quite likely that any GetBuildIndependentHash-style method would also compute different hashes.

In any case, I don't think we're going to add features unrelated to decompilation just to workaround what might be a Roslyn bug.

**Comment by tapika** [comment]
Doubt it's Roslyn bug, as binary code which changes is located outside of IL code, so suspect it's either alignment or full bytes, not relates to executable code.

Also what I have tested is .net framework, and newest features do end up into .net core - so that could be also the reason.

Why computing build independent hash cannot be a part of ilspy API ?

**Comment by tapika** [comment]
What if I'll sketch changes into ilspy, will consider to take my changes as pull request ?

**Comment by christophwille** [comment]
Please feel free to sketch out your idea (after all, open source is about scratching an itch). And by all means, if you feel your approach is up to snuff, do create a PR.

**Comment by siegfriedpammer** [comment]
> Doubt it's Roslyn bug, as binary code which changes is located outside of IL code, so suspect it's either alignment or full bytes, not relates to executable code.
Also what I have tested is .net framework, and newest features do end up into .net core - so that could be also the reason.

In order to help you better and understanding the problem you are trying to solve, I would be interested in seeing an example of a C# program, for which the C# compiler produces a different binary on recompilation despite the use of the `/deterministic` option.

If there exists such a program, it might be worth asking the Roslyn team for help, because the very definition of "deterministic build" is that compiling a source file produces the exact same result on recompilation, regardless of how many times you repeat the process. (For more information on deterministic builds in C# see [here](https://gist.github.com/aelij/b20271f4bd0ab1298e49068b388b54ae))

> Why computing build independent hash cannot be a part of ILSpy API?

Well, we want to focus on decompilation and analysis of the contents of CIL binaries. Comparing executables is a very complex topic, which even warrants having a dedicated tool for it, however we are not interested in building such a tool at the moment.

Thank you for your understanding!

**Comment by tapika** [comment]
> In order to help you better and understanding the problem you are trying to solve, I would 
> be interested in seeing an example of a C# program, for which the C# compiler produces a 
> different binary on recompilation despite the use of the /deterministic option.

I could maybe try to tell why this does not work for me - I'm using .net framework, not
.net core. I suspect that /deterministic flag was introduced for Roslyn compiler, which 
is .net core.

> If there exists such a program, it might be worth asking the Roslyn team for help

Highly doubt they will answer anything - since .net framework is old platform, and they are not planning
any improvement on top of that one.

Like said - I've raised similar ticket for .net core - in here:

https://github.com/dotnet/core/issues/5577

but since nothing is happening in there, I was trying to identify best approach to solve this on my own, 
and then came accross ilspy. ilspy performs everything I need to get this task done - I could also 
try to alter .net core platform itself, but suspect it's easier to go with ilspy-tool, than with framework.

Especially ilspy has a lot of built-in functionality, including producing valid source code.
Using built source code is useful to check whether or not checksum is actually correctly calculated or not.

> Well, we want to focus on decompilation and analysis of the contents of CIL binaries. Comparing 
> executables is a very complex topic, which even warrants having a dedicated tool for it, however we are not interested in building such a tool at the moment.

Doubt it's complex issue, as you might think.

I have already set of assemblies (for .net framework), which do change from build to build - but theit logical content
remains the same. I was able to calculate their checksum using IncrementalHash class for that purpose.
You just need to call it at right phase with right data - that's it.

Attaching my own changes as a quick patch if you want to compare.

But I was also thinking bit deeper to better structure whole system so I can also test the checksum calculation.

Btw, besides checksumming whole assembly - I think it's also possible to checksum individual interfaces in order to 
determine whether or not they are API compatible. I had idea like this already, but need to elaborate it bit deeper later on.
I think checksumming assembly is sufficient for my needs currently, as I can isolate interfaces into separate assemblies if necessary.

It would be nice if you would allow to accept changes which I want to push to ilspy, as otherwise it will be non maintained my own fork of ilspy.

**Comment by tapika** [comment]
[ilspychanges.zip](https://github.com/icsharpcode/ILSpy/files/6072292/ilspychanges.zip)

**Comment by tapika** [comment]
Please note that my own patch are quick and ugly hack. I will try to make better solution than this one.

**Comment by tapika** [comment]
Btw, I've experimented bit more with `<Deterministic>true</Deterministic>`
- originally I saw it was not working, but I have used also `<PathMap>$(EnlistmentRoot)=C:\</PathMap>` parameter
- with this set of parameters dlls always changed.

But today I've discovered that `Deterministic` option work better with either one of these two parameters:
`<PathMap>$([System.IO.Path]::GetFullPath('$(MSBuildThisFileDirectory)'))=./</PathMap>`

or 
`<PathMap>$(SolutionDir)=./</PathMap>`

I've stopped to last one, because suspect path is rooted against intermediate binary output folder, not project or source codes, unlike Microsoft documentation says.

Anyway, I've run these option to compile multiple dll's and managed to get identical output for 2 out of 5 dlls - but 3 dll still did change. (All of dll's are using .net framework)

(Unfortunately cannot provide you source codes of my project as it's not open source code at the moment.)
